src/inmanta/agent/code_manager.py

```diff
@@ -63,71 +62,66 @@ async def get_pip_config(self, environment: uuid.UUID, version: int) -> PipConfi
-        if not resource_types:
-            raise ValueError(f"{self.__class__.__name__}.get_code() expects at least one resource type")
-
-        resource_install_specs: list[ResourceInstallSpec] = []
-        invalid_resources: executor.FailedResources = {}
-        for resource_type in set(resource_types):
-            try:
-                resource_install_specs.append(await self.get_code_for_type(environment, version, resource_type))
-            except CouldNotResolveCode as e:
-                LOGGER.error(
-                    "%s",
-                    e.msg,
-                )
-                invalid_resources[resource_type] = e
-            except Exception as e:
-                LOGGER.error(
-                    "Failed to get source code for %s version=%d",
-                    resource_type,
-                    version,
-                    exc_info=True,
-                )
-                invalid_resources[resource_type] = e
 
-        return resource_install_specs, invalid_resources
+        module_install_specs = []
+        stmt = (
+            select(
+                models.ModulesForAgent.module_name,
+                models.ModulesForAgent.module_version,
+                models.ModulesForAgent.cm_version,
+                models.Module.requirements,
+                models.FilesInModule.file_content_hash,
+            )
+            .join_from(
+                models.ModulesForAgent,
+                models.Module,
+                and_(
+                    models.ModulesForAgent.module_name == models.Module.name,
+                    models.ModulesForAgent.module_version == models.Module.version,
+                    models.ModulesForAgent.environment == models.Module.environment,
+                ),
+            )
+            .join_from(
+                models.Module,
+                models.FilesInModule,
+                and_(
+                    models.Module.name == models.FilesInModule.module_name,
+                    models.Module.version == models.FilesInModule.module_version,
+                    models.Module.environment == models.FilesInModule.environment,
+                ),
+            )
+            .where(
+                models.ModulesForAgent.environment == environment,
+                models.ModulesForAgent.agent_name == agent_name,
+                models.ModulesForAgent.cm_version == model_version,
+            )
+        )
+        async with get_session() as session:
+            result_execute = await session.execute(stmt)
+            for res in result_execute.all():
+                module_install_specs.append(
+                    ModuleInstallSpec(
+                        module_name=res.module_name,
+                        module_version=res.module_version,
+                        model_version=res.cm_version,
+                        blueprint=executor.ExecutorBlueprint(
+                            pip_config=await self.get_pip_config(environment, model_version),
```

Comment:
The pip config could probably be added to the query and perhaps move to one of the new tables

Reply:
> The pip config could probably be added to the query

This part I agree

> and perhaps move to one of the new tables

What is the motivation behind this ? Wouldn't this bring unnecessary duplication i.e. store the same pip config on each row with a similar model version in the `modules_for_agent` table ?